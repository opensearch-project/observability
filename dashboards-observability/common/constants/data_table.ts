/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const GRID_HEADER_COLUMN_MAX_WIDTH = '150px';
export const GRID_PAGE_RANGE_DISPLAY = 5;
export const COLUMN_DEFAULT_MIN_WIDTH = 100;
export const GRID_PAGE_SIZES = [10, 50, 100];
export const ROW_DENSITIES = [
  { icon: 'tableDensityExpanded', height: 55, selected: false },
  { icon: 'tableDensityNormal', height: 45, selected: false },
  { icon: 'tableDensityCompact', height: 35, selected: true },
];

export const HEADER_HEIGHT = 35;
