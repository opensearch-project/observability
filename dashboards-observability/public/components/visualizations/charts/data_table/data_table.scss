.pagination-wrapper {
  display: flex;
  justify-content: space-between;
}
.custom-pagination-container {
  display: inline-flex;
  padding: 10px 5px;
}

.custom-pagination-container li {
  font-size: 14px;
  padding: 0 5px;
  display: flex;
  align-items: flex-end;
  font-weight: 600;
}

.custom-pagination-container .euiIcon:focus {
  background: transparent !important;
}
.custom-pagination-container a:focus {
  animation: none !important;
}
.custom-pagination-container a {
  cursor: pointer;
}

.custom-pagination-container .disabled {
  cursor: default;
}
.custom-pagination-container a {
  color: #343741;
}
.dark-theme a {
  color: #dfe5ef;
}
.custom-pagination-container .selected a {
  color: #1ba9f5;
  text-decoration: underline;
  cursor: default;
}

.custom-overlay {
  display: flex;
  align-items: flex-start;
  padding: 10px;
  position: fixed;
  top: 100px;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 99999 !important;
  overflow-y: scroll;
  overflow-x: hidden;
}

.custom-overlay-light {
  background-color: #fff;
}

.custom-overlay-dark {
  background-color: #16171c;
}
