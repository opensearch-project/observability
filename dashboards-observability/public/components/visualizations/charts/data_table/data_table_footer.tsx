/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import {
  EuiFlexGroup,
  EuiFlexItem,
  EuiButtonEmpty,
  EuiPopover,
  EuiIcon,
  EuiContextMenuPanel,
  EuiContextMenuItem,
} from '@elastic/eui';

// pagination
import ReactPaginate from 'react-paginate';

// theme
import { uiSettingsService } from '../../../../../common/utils';

// constants
import {
  GRID_PAGE_RANGE_DISPLAY,
  GRID_PAGE_SIZES,
} from '../../../../../common/constants/explorer';

export const GridFooter = ({
  onPageSizeChanged,
  pageSize,
  activePage,
  goToPage,
  pageCount,
}: {
  onPageSizeChanged: (val: number) => void;
  goToPage: ({ selected }: { selected: number }) => void;
  pageSize: number;
  activePage: number;
  pageCount: number;
}) => {
  return (
    <EuiFlexGroup
      responsive
      gutterSize="none"
      justifyContent="spaceBetween"
      style={{ paddingTop: '10px', position: 'relative' }}
    >
      <EuiFlexItem grow={false}>
        <PageSizePopover onPageSizeChanged={onPageSizeChanged} pageSize={pageSize} />
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <ReactPaginate
          containerClassName={
            uiSettingsService.get('theme:darkMode')
              ? `custom-pagination-container dark-theme`
              : `custom-pagination-container`
          }
          breakLabel="..."
          nextLabel={<EuiIcon type="arrowRight" />}
          forcePage={activePage}
          onPageChange={goToPage}
          pageRangeDisplayed={GRID_PAGE_RANGE_DISPLAY}
          pageCount={pageCount}
          previousLabel={<EuiIcon type="arrowLeft" />}
          renderOnZeroPageCount={null}
        />
      </EuiFlexItem>
    </EuiFlexGroup>
  );
};

export const PageSizePopover = ({
  onPageSizeChanged,
  pageSize,
}: {
  onPageSizeChanged: (size: number) => void;
  pageSize: number;
}) => {
  const [isPopoverOpen, setIsPopoverOpen] = useState<boolean>(false);
  const onButtonClick = () => setIsPopoverOpen((isPopoverOpen) => !isPopoverOpen);
  const closePopover = () => setIsPopoverOpen(false);

  const button = (
    <EuiButtonEmpty
      iconSize="s"
      color="text"
      size="s"
      aria-label="Next"
      iconType="arrowDown"
      iconSide="right"
      onClick={onButtonClick}
    >
      Rows per page: {pageSize}
    </EuiButtonEmpty>
  );

  const items = () => {
    return GRID_PAGE_SIZES.map((i: number) => (
      <EuiContextMenuItem
        layoutAlign="bottom"
        key={`${i}_rows`}
        icon={i === pageSize ? 'check' : 'empty'}
        onClick={() => {
          onPageSizeChanged(i);
          closePopover();
        }}
      >
        {i} rows
      </EuiContextMenuItem>
    ));
  };

  return (
    <EuiPopover
      id="singlePanel"
      button={button}
      isOpen={isPopoverOpen}
      closePopover={closePopover}
      panelPaddingSize="none"
      anchorPosition="downLeft"
    >
      <EuiContextMenuPanel items={items()} />
    </EuiPopover>
  );
};
