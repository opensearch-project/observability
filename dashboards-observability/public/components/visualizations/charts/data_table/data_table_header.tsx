/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState, ReactChildren, ReactChild } from 'react';
import {
  EuiButtonIcon,
  EuiFlexGroup,
  EuiFlexItem,
  EuiButtonEmpty,
  EuiPopover,
  EuiSwitch,
  EuiIcon,
} from '@elastic/eui';

// theme
import { uiSettingsService } from '../../../../../common/utils';

// constants
import {
  GRID_HEADER_COLUMN_MAX_WIDTH,
  ROW_DENSITIES,
} from '../../../../../common/constants/explorer';

export interface RowConfigType {
  icon: string;
  height: number;
  selected: boolean;
}

export const CustomOverlay = ({
  children,
}: {
  children: ReactChild | ReactChild[] | ReactChildren | ReactChildren[];
}) => {
  return (
    <div
      className={
        uiSettingsService.get('theme:darkMode')
          ? 'custom-overlay custom-overlay-dark'
          : 'custom-overlay custom-overlay-light'
      }
    >
      {children}
    </div>
  );
};

export const GridHeader = ({
  isFullScreen,
  setIsFullScreenHandler,
  selectedRowDensity,
  selectDensityHandler,
  columnVisiblityHandler,
  columns,
  columnVisibility,
}: {
  isFullScreen: boolean;
  setIsFullScreenHandler: (v: boolean) => void;
  selectedRowDensity: any;
  selectDensityHandler: (v: RowConfigType) => void;
  columnVisiblityHandler: (visible: boolean, field: string) => void;
  columns: any;
  columnVisibility: any;
}) => {
  return (
    <EuiFlexGroup
      responsive
      gutterSize="none"
      justifyContent="flexStart"
      style={{ position: 'relative' }}
    >
      <EuiFlexItem style={{ maxWidth: GRID_HEADER_COLUMN_MAX_WIDTH }}>
        <DensityPopover
          selectedDensity={selectedRowDensity}
          selectDensityHandler={selectDensityHandler}
        />
      </EuiFlexItem>
      <EuiFlexItem style={{ maxWidth: GRID_HEADER_COLUMN_MAX_WIDTH }}>
        <ColumnVisiblityPopover
          columnVisiblityHandler={columnVisiblityHandler}
          columns={columns}
          columnVisibility={columnVisibility}
        />
      </EuiFlexItem>
      <EuiFlexItem style={{ maxWidth: GRID_HEADER_COLUMN_MAX_WIDTH }}>
        <EuiButtonEmpty
          iconSize="s"
          color="text"
          aria-label="Next"
          iconType="fullScreen"
          onClick={() => setIsFullScreenHandler(true)}
        >
          Full screen
        </EuiButtonEmpty>
      </EuiFlexItem>
      {isFullScreen && (
        <EuiIcon
          type="cross"
          onClick={() => setIsFullScreenHandler(false)}
          style={{ position: 'absolute', right: 20, cursor: 'pointer', top: 20 }}
        />
      )}
    </EuiFlexGroup>
  );
};

export const DensityPopover = ({
  selectDensityHandler,
  selectedDensity,
}: {
  selectedDensity: any;
  selectDensityHandler: (data: RowConfigType) => void;
}) => {
  const [isPopoverOpen, setIsPopoverOpen] = useState<boolean>(false);
  const onButtonClick = () => setIsPopoverOpen((isPopoverOpen) => !isPopoverOpen);
  const closePopover = () => setIsPopoverOpen(false);

  const button = (
    <EuiButtonEmpty
      iconSize="s"
      color="text"
      aria-label="Next"
      iconType={selectedDensity.icon}
      onClick={onButtonClick}
    >
      Density
    </EuiButtonEmpty>
  );

  return (
    <EuiPopover button={button} isOpen={isPopoverOpen} closePopover={closePopover}>
      <EuiFlexGroup responsive={false} gutterSize="s" alignItems="center">
        {ROW_DENSITIES.map((i: RowConfigType, index: number) => (
          <EuiFlexItem key={index} grow={false}>
            <EuiButtonIcon
              onClick={() => selectDensityHandler(i)}
              display={selectedDensity.icon === i.icon ? 'fill' : 'base'}
              iconType={i.icon}
              aria-label="Next"
            />
          </EuiFlexItem>
        ))}
      </EuiFlexGroup>
    </EuiPopover>
  );
};

export const ColumnVisiblityPopover = ({
  columnVisibility,
  columns,
  columnVisiblityHandler,
}: {
  columns: any;
  columnVisibility: any;
  columnVisiblityHandler: (visible: boolean, field: string) => void;
}) => {
  const [isPopoverOpen, setIsPopoverOpen] = useState<boolean>(false);
  const onButtonClick = () => setIsPopoverOpen((isPopoverOpen) => !isPopoverOpen);
  const closePopover = () => setIsPopoverOpen(false);

  const button = (
    <EuiButtonEmpty
      iconSize="s"
      color="text"
      aria-label="Next"
      iconType="listAdd"
      onClick={onButtonClick}
    >
      Columns
    </EuiButtonEmpty>
  );

  return (
    <EuiPopover button={button} isOpen={isPopoverOpen} closePopover={closePopover}>
      <EuiFlexGroup gutterSize="s" direction="column">
        {columns.map((i: any, index: number) => {
          return (
            <EuiFlexItem key={index} grow={false}>
              <EuiSwitch
                label={i.field}
                checked={!columnVisibility.includes(i.field)}
                onChange={(e) => {
                  columnVisiblityHandler(e.target.checked, i.field);
                }}
                compressed
              />
            </EuiFlexItem>
          );
        })}
      </EuiFlexGroup>
    </EuiPopover>
  );
};
