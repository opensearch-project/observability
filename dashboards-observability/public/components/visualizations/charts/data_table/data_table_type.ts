/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DataTable } from './data_table';
import { getPlotlySharedConfigs, getPlotlyCategory } from '../shared/shared_configs';
import { LensIconChartDatatable } from '../../assets/chart_datatable';
import { ConfigEditor } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/json_editor';
import { ConfigAvailability } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_availability';
import { VizDataPanel } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/default_vis_editor';
import {
  InputFieldItem,
  SwitchButton,
  ConfigChartOptions,
  ButtonGroupItem,
} from './../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls';

const sharedConfigs = getPlotlySharedConfigs();
const VIS_CATEGORY = getPlotlyCategory();
const COLUMN_WIDTH = 150;

export const createDatatableTypeDefinition = (params: any = {}) => ({
  name: 'data_table',
  type: 'data_table',
  id: 'data_table',
  label: 'Table View',
  fulllabel: 'Table View',
  icontype: 'visTable',
  category: VIS_CATEGORY.BASICS,
  icon: LensIconChartDatatable,
  showtableheader: true,
  enablepagination: true,
  colunmfilter: false,
  columnalignment: 'leftAligned',
  columnwidth: COLUMN_WIDTH,
  editorconfig: {
    panelTabs: [
      {
        id: 'data-panel',
        name: 'Data',
        mapTo: 'dataConfig',
        editor: VizDataPanel,
        sections: [
          {
            id: 'chart-styles',
            name: 'Chart styles',
            editor: ConfigChartOptions,
            mapTo: 'chartStyles',
            schemas: [
              {
                title: 'Show table header',
                name: 'Show table header',
                component: SwitchButton,
                mapTo: 'showTableHeader',
                eleType: 'switchButton',
                currentValue: true,
              },
              {
                title: 'Enable pagination',
                name: 'Enable pagination',
                component: SwitchButton,
                mapTo: 'enablePagination',
                eleType: 'switchButton',
                currentValue: true,
              },
              {
                name: 'Column alignment',
                component: ButtonGroupItem,
                mapTo: 'columnAlignment',
                eleType: 'buttons',
                props: {
                  options: [
                    { name: 'Left', id: 'leftAligned' },
                    { name: 'Right', id: 'rightAligned' },
                  ],
                  defaultSelections: [{ name: 'Left', id: 'leftAligned' }],
                },
              },
              {
                title: 'Column Width',
                name: 'Column Width',
                component: InputFieldItem,
                mapTo: 'columnWidth',
                eleType: 'input',
                currentValue: 150,
              },
              {
                title: 'Column filter',
                name: 'Column filter',
                component: SwitchButton,
                mapTo: 'colunmFilter',
                eleType: 'switchButton',
                currentValue: false,
              },
            ],
          },
        ],
      },
    ],
  },
  visconfig: {
    layout: {
      ...sharedConfigs.layout,
    },
    config: {
      ...sharedConfigs.config,
    },
    isUniColor: false,
  },
  component: DataTable,
});
