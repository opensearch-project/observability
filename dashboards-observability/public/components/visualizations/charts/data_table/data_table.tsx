/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState, useMemo, useCallback, useRef, useEffect } from 'react';
import { EuiFlexGroup, EuiFlexItem } from '@elastic/eui';

// ag-data-grid
import { AgGridReact } from 'ag-grid-react';
import 'ag-grid-community/dist/styles/ag-grid.css';
import 'ag-grid-community/dist/styles/ag-theme-alpine.css';

// grid elements
import { CustomOverlay, RowConfigType, GridHeader } from './data_table_header';
import { GridFooter } from './data_table_footer';

// constants
import { COLUMN_DEFAULT_MIN_WIDTH, HEADER_HEIGHT } from '../../../../../common/constants/explorer';
import { IVisualizationContainerProps, IField } from '../../../../../common/types/explorer';
import 'ag-grid-community/dist/styles/ag-theme-alpine-dark.css';

// styles
import './data_table.scss';

const doubleValueGetter = (params) => {
  return params.data[params.column.colId];
};

export const DataTable = ({ visualizations }: any) => {
  const {
    data: {
      rawVizData: {
        jsonData = [],
        metadata: { fields = [] } = {},
      } = {},
      userConfigs: { dataConfig: { chartStyles = {} } = {} } = {},
    } = {}
  }: IVisualizationContainerProps = visualizations;

  const enablePagination =
    typeof chartStyles.enablePagination !== 'undefined'
      ? chartStyles.enablePagination
      : visualizations.vis.enablepagination;

  const showTableHeader =
    typeof chartStyles.showTableHeader !== 'undefined'
      ? chartStyles.showTableHeader
      : visualizations.vis.showtableheader;

  const colunmFilter =
    typeof chartStyles.colunmFilter !== 'undefined'
      ? chartStyles.colunmFilter
      : visualizations.vis.colunmfilter;

  const columnAlignment = chartStyles.columnAlignment || visualizations.vis.columnalignment;

  const columnWidth =
    typeof chartStyles.columnWidth !== 'undefined'
      ? chartStyles.columnWidth
      : visualizations.vis.columnwidth;

  useEffect(() => {
    document.addEventListener('keydown', hideGridFullScreenHandler);
    return () => {
      document.removeEventListener('keydown', hideGridFullScreenHandler);
    };
  }, []);

  // rows and columns
  const rawData = [...jsonData];

  const columns = useMemo(
    () =>
      fields.map((field: IField) => {
        return {
          lockVisible: true,
          columnsMenuParams: {
            suppressColumnFilter: true,
            suppressColumnSelectAll: true,
            suppressColumnExpandAll: true,
          },
          headerName: field.name,
          field: field.name,
          colId: field.name,
          ...(field.type === 'double' && {
            valueGetter: doubleValueGetter,
          }),
        };
      }),
    [fields]
  );

  // ag-grid-react bindings
  const gridRef = useRef<any | undefined>();
  const gridRefFullScreen = useRef<any | undefined>();
  const [pageSize, setPageSize] = useState<number>(10);
  const [columnVisibility, setColumnVisibility] = useState<string[]>([]);
  const [isFullScreen, setIsFullScreen] = useState<boolean>(false);
  const [selectedRowDensity, setSelectedRowDensity] = useState<RowConfigType>({
    icon: 'tableDensityCompact',
    height: 35,
    selected: true,
  });
  // pagination
  const [activePage, setActivePage] = useState<number>(0);
  const pageCount = Math.ceil(rawData.length / pageSize);

  const defaultColDef = useMemo(() => {
    return {
      sortable: true,
      resizable: true,
      filter: colunmFilter,
      flex: 1,
      suppressMenu: false,
      minWidth: COLUMN_DEFAULT_MIN_WIDTH,
      headerHeight: 400,
      type: columnAlignment,
    };
  }, [colunmFilter, columnAlignment]);

  useEffect(() => {
    if (!chartStyles.columnWidth) {
      gridRef?.current?.api?.sizeColumnsToFit();
    } else {
      columns.forEach((col: any) =>
        gridRef?.current?.columnApi?.setColumnWidth(col.field, Number(columnWidth))
      );
    }
  }, [columnWidth, columns]);

  const onPageSizeChanged = useCallback(
    (val: number) => {
      setPageSize(val);
      gridRef.current.api.paginationSetPageSize(val);
      setActivePage(0);
      gridRef.current.api.paginationGoToPage(0);
      if (isFullScreen) {
        gridRefFullScreen.current.api.paginationSetPageSize(val);
        gridRefFullScreen.current.api.paginationGoToPage(0);
      }
    },
    [isFullScreen]
  );

  const selectDensityHandler = useCallback((value: RowConfigType) => {
    setSelectedRowDensity({ ...value });
    gridRef.current.api.forEachNode((rowNode) => {
      if (rowNode.data) {
        rowNode.setRowHeight(value.height);
      }
    });
    gridRef.current.api.onRowHeightChanged();
  }, []);

  const columnVisiblityHandler = useCallback((visible: boolean, field: string) => {
    const isExisting = columnVisibility.findIndex((column) => column === field);
    if (visible) {
      if (isExisting > -1) {
        columnVisibility.splice(isExisting, 1);
        gridRef?.current?.columnApi?.setColumnsVisible([field], true);
      }
    } else {
      columnVisibility.push(field);
      gridRef?.current?.columnApi?.setColumnsVisible([field], false);
    }
    setColumnVisibility([...columnVisibility]);
  }, []);

  const goToPage = ({ selected }: { selected: number }) => {
    setActivePage(selected);
    gridRef.current.api.paginationGoToPage(selected);
    if (isFullScreen) {
      gridRefFullScreen.current.api.paginationGoToPage(selected);
    }
  };

  const setIsFullScreenHandler = (val: boolean) => {
    setIsFullScreen(val);
  };

  const hideGridFullScreenHandler = (e: any) => {
    if (e.key === 'Escape') {
      setIsFullScreen(false);
    }
  };

  return (
    <>
      {showTableHeader && (
        <GridHeader
          isFullScreen={isFullScreen}
          setIsFullScreenHandler={setIsFullScreenHandler}
          selectedRowDensity={selectedRowDensity}
          selectDensityHandler={selectDensityHandler}
          columnVisiblityHandler={columnVisiblityHandler}
          columns={columns}
          columnVisibility={columnVisibility}
        />
      )}
      <AgGridReact
        ref={gridRef}
        rowData={rawData}
        columnDefs={columns}
        defaultColDef={defaultColDef}
        domLayout={'autoHeight'}
        animateRows
        pagination={enablePagination}
        paginationPageSize={pageSize}
        suppressPaginationPanel
        rowHeight={selectedRowDensity.height}
        onGridReady={() => {
          gridRef?.current?.api.setHeaderHeight(HEADER_HEIGHT);
        }}
        suppressFieldDotNotation // added for key contains dot operator
      />
      {enablePagination && (
        <GridFooter
          onPageSizeChanged={onPageSizeChanged}
          pageSize={pageSize}
          activePage={activePage}
          goToPage={goToPage}
          pageCount={pageCount}
        />
      )}
      {isFullScreen && (
        <CustomOverlay>
          <EuiFlexGroup direction="column">
            <EuiFlexItem>
              <AgGridReact
                ref={gridRefFullScreen}
                rowData={rawData}
                columnDefs={columns}
                defaultColDef={defaultColDef}
                domLayout="autoHeight"
                animateRows
                pagination
                paginationPageSize={pageSize}
                suppressPaginationPanel
                rowHeight={selectedRowDensity.height}
                onGridReady={() => {
                  gridRefFullScreen?.current?.api.setHeaderHeight(HEADER_HEIGHT);
                }}
                // added for key contains dot operator
                suppressFieldDotNotation
              />
            </EuiFlexItem>
            <EuiFlexItem>
              <GridFooter
                onPageSizeChanged={onPageSizeChanged}
                pageSize={pageSize}
                activePage={activePage}
                goToPage={goToPage}
                pageCount={pageCount}
              />
            </EuiFlexItem>
          </EuiFlexGroup>
        </CustomOverlay>
      )}
    </>
  );
};
