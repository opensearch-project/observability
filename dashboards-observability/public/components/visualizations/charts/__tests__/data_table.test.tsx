/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { DataTable } from '../data_table/data_table';
import { TEST_VISUALIZATIONS_DATA } from '../../../../../test/event_analytics_constants';
import { AgGridReact } from 'ag-grid-react';

describe('Data table component', () => {
  configure({ adapter: new Adapter() });

  it('Renders data table component', async () => {
    const gridWrapper = shallow(<DataTable visualizations={TEST_VISUALIZATIONS_DATA} />);
    const agGridReactObj = gridWrapper.find(AgGridReact);
    agGridReactObj.simulate('gridReady');
    expect(agGridReactObj).toBeTruthy();
    await waitFor(() => {
      expect(gridWrapper).toMatchSnapshot();
    });
  });
});
