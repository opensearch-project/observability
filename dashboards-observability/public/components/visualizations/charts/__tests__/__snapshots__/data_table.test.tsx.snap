// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data table component Renders data table component 1`] = `
<DataTable
  visualizations={
    Object {
      "data": Object {
        "appData": Object {
          "fromApp": false,
        },
        "defaultAxes": Object {},
        "indexFields": Object {
          "availableFields": Array [
            Object {
              "name": "agent",
              "type": "string",
            },
            Object {
              "name": "bytes",
              "type": "long",
            },
            Object {
              "name": "clientip",
              "type": "ip",
            },
            Object {
              "name": "event",
              "type": "struct",
            },
            Object {
              "name": "extension",
              "type": "string",
            },
            Object {
              "name": "geo",
              "type": "struct",
            },
            Object {
              "name": "host",
              "type": "string",
            },
            Object {
              "name": "index",
              "type": "string",
            },
            Object {
              "name": "ip",
              "type": "ip",
            },
            Object {
              "name": "machine",
              "type": "struct",
            },
            Object {
              "name": "memory",
              "type": "double",
            },
            Object {
              "name": "message",
              "type": "string",
            },
            Object {
              "name": "phpmemory",
              "type": "long",
            },
            Object {
              "name": "referer",
              "type": "string",
            },
            Object {
              "name": "request",
              "type": "string",
            },
            Object {
              "name": "response",
              "type": "string",
            },
            Object {
              "name": "tags",
              "type": "string",
            },
            Object {
              "name": "timestamp",
              "type": "timestamp",
            },
            Object {
              "name": "url",
              "type": "string",
            },
            Object {
              "name": "utc_time",
              "type": "timestamp",
            },
          ],
          "queriedFields": Array [
            Object {
              "name": "double_per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "host",
              "type": "text",
            },
            Object {
              "name": "ip_count",
              "type": "integer",
            },
            Object {
              "name": "per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "resp_code",
              "type": "text",
            },
            Object {
              "name": "sum_bytes",
              "type": "long",
            },
          ],
          "selectedFields": Array [],
          "unselectedFields": Array [],
        },
        "query": Object {},
        "rawVizData": Object {
          "data": Object {
            "count()": Array [
              154,
              1753,
              116,
              468,
              1964,
              219,
            ],
            "tags": Array [
              "error",
              "info",
              "login",
              "security",
              "success",
              "warning",
            ],
          },
          "jsonData": Array [
            Object {
              "count()": 154,
              "tags": "error",
            },
            Object {
              "count()": 1753,
              "tags": "info",
            },
            Object {
              "count()": 116,
              "tags": "login",
            },
            Object {
              "count()": 468,
              "tags": "security",
            },
            Object {
              "count()": 1964,
              "tags": "success",
            },
            Object {
              "count()": 219,
              "tags": "warning",
            },
          ],
          "metadata": Object {
            "fields": Array [
              Object {
                "name": "count()",
                "type": "integer",
              },
              Object {
                "name": "tags",
                "type": "text",
              },
            ],
            "size": 6,
            "status": 200,
          },
        },
        "userConfigs": Object {},
      },
      "vis": Object {
        "category": "Visualizations",
        "categoryAxis": "xaxis",
        "component": [Function],
        "editorConfig": Object {
          "panelTabs": Array [
            Object {
              "editor": [Function],
              "id": "data-panel",
              "mapTo": "dataConfig",
              "name": "Data",
              "sections": Array [
                Object {
                  "editor": [Function],
                  "id": "value_options",
                  "mapTo": "valueOptions",
                  "name": "Value options",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "isSingleSelection": false,
                      "mapTo": "xaxis",
                      "name": "X-axis",
                    },
                    Object {
                      "component": null,
                      "isSingleSelection": false,
                      "mapTo": "yaxis",
                      "name": "Y-axis",
                    },
                  ],
                },
                Object {
                  "editor": [Function],
                  "id": "chart_options",
                  "mapTo": "chartOptions",
                  "name": "Chart options",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "isSingleSelection": true,
                      "mapTo": "orientation",
                      "name": "Orientation",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "name": "Vertical",
                            "orientationId": "v",
                          },
                        ],
                        "dropdownList": Array [
                          Object {
                            "name": "Vertical",
                            "orientationId": "v",
                          },
                          Object {
                            "name": "Horizontal",
                            "orientationId": "h",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": null,
                      "isSingleSelection": true,
                      "mapTo": "mode",
                      "name": "Mode",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "modeId": "group",
                            "name": "Group",
                          },
                        ],
                        "dropdownList": Array [
                          Object {
                            "modeId": "group",
                            "name": "Group",
                          },
                          Object {
                            "modeId": "stack",
                            "name": "Stack",
                          },
                        ],
                      },
                    },
                  ],
                },
              ],
            },
            Object {
              "content": Array [],
              "editor": [Function],
              "id": "style-panel",
              "mapTo": "layoutConfig",
              "name": "Layout",
            },
            Object {
              "editor": [Function],
              "id": "availability-panel",
              "mapTo": "availabilityConfig",
              "name": "Availability",
            },
          ],
        },
        "fullLabel": "Bar",
        "icon": [Function],
        "iconType": "visBarVerticalStacked",
        "id": "bar",
        "label": "Bar",
        "name": "bar",
        "orientation": "v",
        "selection": Object {
          "dataLoss": "nothing",
        },
        "seriesAxis": "yaxis",
        "type": "bar",
        "visConfig": Object {
          "config": Object {
            "displaylogo": false,
            "responsive": true,
          },
          "isUniColor": false,
          "layout": Object {
            "height": 500,
            "legend": Object {
              "orientation": "v",
              "traceorder": "normal",
            },
            "margin": Object {
              "b": 30,
              "l": 60,
              "pad": 0,
              "r": 30,
              "t": 50,
            },
            "showlegend": true,
          },
        },
      },
    }
  }
>
  <EuiDataGrid
    aria-label="viz data table"
    columnVisibility={
      Object {
        "setVisibleColumns": [Function],
        "visibleColumns": Array [
          "count()",
          "tags",
        ],
      }
    }
    columns={
      Array [
        Object {
          "id": "count()",
        },
        Object {
          "id": "tags",
        },
      ]
    }
    data-test-subj="workspace__dataTable"
    pagination={
      Object {
        "onChangeItemsPerPage": [Function],
        "onChangePage": [Function],
        "pageIndex": 0,
        "pageSize": 10,
        "pageSizeOptions": Array [
          10,
          50,
          100,
        ],
      }
    }
    renderCellValue={[Function]}
    rowCount={6}
    sorting={
      Object {
        "columns": Array [],
        "onSort": [Function],
      }
    }
  >
    <EuiFocusTrap
      className="euiDataGrid__focusWrap"
      clickOutsideDisables={false}
      disabled={true}
      noIsolation={true}
      returnFocus={true}
      scrollLock={false}
    >
      <ForwardRef
        className="euiDataGrid__focusWrap"
        enabled={false}
        noIsolation={true}
        onClickOutside={[Function]}
        returnFocus={true}
        scrollLock={false}
      >
        <ForwardRef
          className="euiDataGrid__focusWrap"
          enabled={false}
          noIsolation={true}
          onClickOutside={[Function]}
          returnFocus={true}
          scrollLock={false}
          sideCar={[Function]}
        >
          <ForwardRef(FocusLockUI)
            as={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "classNames": Object {
                  "fullWidth": "width-before-scroll-bar",
                  "zeroRight": "right-scroll-bar-position",
                },
                "defaultProps": Object {
                  "enabled": true,
                  "inert": false,
                  "removeScrollBar": true,
                },
                "render": [Function],
              }
            }
            autoFocus={true}
            className="euiDataGrid__focusWrap"
            crossFrame={true}
            disabled={true}
            lockProps={
              Object {
                "allowPinchZoom": undefined,
                "as": undefined,
                "enabled": false,
                "inert": undefined,
                "shards": undefined,
                "sideCar": [Function],
                "style": undefined,
              }
            }
            noFocusGuards={false}
            persistentFocus={false}
            returnFocus={true}
            sideCar={[Function]}
          >
            <div
              data-focus-guard={true}
              key="guard-first"
              style={
                Object {
                  "height": "0px",
                  "left": "1px",
                  "overflow": "hidden",
                  "padding": 0,
                  "position": "fixed",
                  "top": "1px",
                  "width": "1px",
                }
              }
              tabIndex={-1}
            />
            <ForwardRef
              className="euiDataGrid__focusWrap"
              data-focus-lock-disabled="disabled"
              enabled={false}
              inert={false}
              onBlur={[Function]}
              onFocus={[Function]}
              removeScrollBar={true}
              sideCar={[Function]}
            >
              <div
                className="euiDataGrid__focusWrap"
                data-focus-lock-disabled="disabled"
                onBlur={[Function]}
                onFocus={[Function]}
                onScrollCapture={[Function]}
                onTouchMoveCapture={[Function]}
                onWheelCapture={[Function]}
              >
                <div
                  className="euiDataGrid euiDataGrid--bordersAll euiDataGrid--headerShade euiDataGrid--footerOverline euiDataGrid--rowHoverHighlight euiDataGrid--stickyFooter"
                  data-test-subj="workspace__dataTable"
                  onKeyDown={[Function]}
                  style={
                    Object {
                      "height": undefined,
                      "width": undefined,
                    }
                  }
                />
              </div>
            </ForwardRef>
            <div
              data-focus-guard={true}
              style={
                Object {
                  "height": "0px",
                  "left": "1px",
                  "overflow": "hidden",
                  "padding": 0,
                  "position": "fixed",
                  "top": "1px",
                  "width": "1px",
                }
              }
              tabIndex={-1}
            />
          </ForwardRef(FocusLockUI)>
        </ForwardRef>
      </ForwardRef>
    </EuiFocusTrap>
  </EuiDataGrid>
</DataTable>
`;
